### DISCUSSION TITLE
Emit warning for shadowed tests

### DISCUSSION THREAD

**Discussion by SullivanMuse** [pytest-dev/pytest#12794] [Ideas] [answered]
Functions that are redefined are not detected by pytest. It would be nice if pytest would emit a warning (or an error!) in these cases, or simply allow multiple tests with the same name.

`main.py`:

```py
def test_bool():
    assert False

def test_bool():
    assert False

def test_bool():
    assert True
```

Output of `pytest main.py`:

```
============================== test session starts ==============================
platform linux -- Python 3.10.12, pytest-8.3.2, pluggy-1.5.0
rootdir: /path/to/parent/dir
plugins: xdoctest-1.2.0, pretty-1.2.0
collected 1 item                                                                

main.py .                                                                 [100%]
Results (0.01s):
         1 passed
```

Posting here rather than feature suggestions because I'm not sure if there is some flag or plugin that enables this.

**Comment by The-Compiler** [comment, marked as answer]
The problem is that pytest cannot trivially do this: Outside of inspecting the [ast](https://en.wikipedia.org/wiki/Abstract_syntax_tree), there is no way for pytest to find out about this - at runtime, pytest only sees the last function, because a function definition is basically just an assignment.

However, there are already plenty of linters that *do* detect this via the AST (ruff, mypy, flake8, pylint, your IDE probably, ...). I don't think there is much value in pytest adding all this complexity (even though I agree this is a common pitfall), since I'd expect a typical Python development environment to already take care of this.

**Reply by RonnyPfannschmidt** [reply]
In case of ast rewrite we could add a warning, it might be a nice deep dive task for someone new
